src/big/big_int.cr

```diff
@@ -621,92 +633,114 @@ struct BigInt < Int
     to_i32
   end
 
-  def to_i8 : Int8
-    to_i32.to_i8
+  def to_i! : Int32
+    to_i32!
   end
 
-  def to_i16 : Int16
-    to_i32.to_i16
+  def to_u : UInt32
+    to_u32
   end
 
-  def to_i32 : Int32
-    LibGMP.get_si(self).to_i32
+  def to_u! : UInt32
+    to_u32!
   end
 
-  def to_i64 : Int64
-    if LibGMP::Long::MIN <= self <= LibGMP::Long::MAX
-      LibGMP.get_si(self).to_i64
-    else
-      to_s.to_i64 { raise OverflowError.new }
+  {% for n in [8, 16, 32, 64, 128] %}
+    def to_i{{n}} : Int{{n}}
+      \{% if Int{{n}} == LibGMP::Long %}
+        LibGMP.fits_slong_p(self) != 0 ? LibGMP.get_si(self) : raise OverflowError.new
+      \{% elsif Int{{n}}::MAX.is_a?(NumberLiteral) && Int{{n}}::MAX < LibGMP::Long::MAX %}
+        LibGMP::Long.new(self).to_i{{n}}
+      \{% else %}
+        to_primitive_i(Int{{n}})
+      \{% end %}
     end
-  end
 
-  def to_i!
-    to_i32!
-  end
-
-  def to_i8! : Int8
-    LibGMP.get_si(self).to_i8!
-  end
+    def to_u{{n}} : UInt{{n}}
+      \{% if Int{{n}} == LibGMP::Long %}
+        LibGMP.fits_ulong_p(self) != 0 ? LibGMP.get_ui(self) : raise OverflowError.new
+      \{% elsif Int{{n}}::MAX.is_a?(NumberLiteral) && Int{{n}}::MAX < LibGMP::Long::MAX %}
```

Comment:
Should this be `ULong`?
```suggestion
      \{% if Int{{n}} == LibGMP::ULong %}
        LibGMP.fits_ulong_p(self) != 0 ? LibGMP.get_ui(self) : raise OverflowError.new
      \{% elsif Int{{n}}::MAX.is_a?(NumberLiteral) && Int{{n}}::MAX < LibGMP::ULong::MAX %}
```

Reply:
It is indeed checking `Int*` against `LibGMP::Long`, botg signed, otherwise they'll never match

Reply:
Right, the other part needs to change, too. That sounds like it would make sense, or do you think otherwise?
I'm unclear what "technically" is supposed to mean.